tests/trainer/test_trainer.py

```diff
@@ -938,8 +938,8 @@ def training_step_and_backward(split_batch, batch_idx, opt_idx, optimizer, hidde
         parameters = model.parameters()
         grad_max_list = [torch.max(p.grad.detach().abs()) for p in parameters]
         grad_max = torch.max(torch.stack(grad_max_list))
-        assert round(grad_max.item(), 6) <= grad_clip_val, \
-            f"Gradient max value {grad_max} > grad_clip_val {grad_clip_val} ."
+        assert abs(round(grad_max.item(), 6) - grad_clip_val) < 1e-6, \
+            f"Gradient max value {grad_max} != grad_clip_val {grad_clip_val} ."
```

Comment:
Yeah, I'm aware. I'd argue that the possibility of that happens is really low (since the threshold is 1e-5). At least I haven't encountered one in my local testing. If you want to make it even lower, we can set the threshold to 1e-10 or 1e-13 (within the fp16 range).

This is the way I can think of to distinguish between clipping by norm and clipping by value. I'm open to better ideas, of course.

Reply:
If you really want to prevent that false positive case to happen, we can add an if statement before that assertion to make sure the minimum gradient is larger than the threshold (this might create some false negatives, though).

Reply:
Can you set the seed, so the test is reproducible ?